### PR TITLE
Add ThreadPool constructor that accepts name as a String

### LINKED ISSUE #26028: BulkProcessor can't be used with empty settings in the high level client
According to [the documentation](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/5.6/java-rest-high-document-bulk.html#java-rest-high-document-bulk-processor), you "only" need to write the following code to create a Rest Bulk Processor:

```java
        ThreadPool threadPool = new ThreadPool(Settings.EMPTY);
        BulkProcessor.Listener listener = new BulkProcessor.Listener() {
            @Override public void beforeBulk(long executionId, BulkRequest request) { }
            @Override public void afterBulk(long executionId, BulkRequest request, BulkResponse response) { }
            @Override public void afterBulk(long executionId, BulkRequest request, Throwable failure) { }
        };

        bulkProcessorDoc = new BulkProcessor.Builder(client.getHlClient()::bulkAsync, listener, threadPool).build();
```

This is actually failing with empty settings as the `ThreadPool` class [is waiting for a `node.name` setting](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L166):

```java
        assert Node.NODE_NAME_SETTING.exists(settings);
```

I think we should either remove the need for the node name when configuring a Client ThreadPool or to document that this is needed for now:

```java
        ThreadPool threadPool = new ThreadPool(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "high-level-client").build());
```

I wonder though if the High Level Rest Client itself should not provide OOTB a default ThreadPool with a name like `high-level-client`.

Thoughts?

### PR DESCRIPTION
**Pull request by javanna**
`BulkProcessor` needs a `ThreadPool` to be created, which needs settings that must contain a `node.name` setting. In a client, such setting makes little sense. This commit allows to create a `ThreadPool` by providing its name separately as a `String`, which will internally be added to the settings. `TestThreadPool` can be removed and replaced with the ordinary `ThreadPool`.

Closes #26028

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 1]
```diff
@@ -160,6 +160,14 @@ public static ThreadPoolType fromType(String type) {
     public static Setting<TimeValue> ESTIMATED_TIME_INTERVAL_SETTING =
         Setting.timeSetting("thread_pool.estimated_time_interval", TimeValue.timeValueMillis(200), Setting.Property.NodeScope);
 
+    public ThreadPool(String name, final ExecutorBuilder<?>... customBuilders) {
+        this(name, Settings.EMPTY, customBuilders);
+    }
+
+    public ThreadPool(String name, Settings settings, final ExecutorBuilder<?>... customBuilders) {
+        this(Settings.builder().put(settings).put(Node.NODE_NAME_SETTING.getKey(), name).build(), customBuilders);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: I think we should fail if `settings` contains `node.name` setting already?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one comment.

**Review by nik9000 — COMMENTED** [review]
I think it'd be cleaner to pass the name where it is needed rather than add it to the Settings object. I know that is a bigger change but stuffing it in a NODE_NAME_SETTING sees a bit hacky to me.

**Review by tlrx — COMMENTED** [review]
I tend to agree with Nik, I find it a bit hacky too.

**Comment by javanna** [comment]
@nik9000 @tlrx the `node.name` was already part of the settings, I only added a new constructor that takes it as a separate argument to work around the check around it, as it makes little sense to require `node.name` setting on client code. Are you suggesting to remove it from there? I wouldn't want this change to affect that much the server code.

**Comment by tlrx** [comment]
@javanna Looking at this again, I think that I'm ok with the additional ctor you added. What I'm not comfortable with is that the node name (or thread pool name) is extracted from the settings lower in the code path, making the node.name setting a mandatory setting but it's not exposed as such by the current ctor `ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBuilders)`. I think we could change that in core but maybe in another pull request.

**Comment by javanna** [comment]
I made the change that @jasontedor requested and added a test. Ready for another review round.

**Comment by jasontedor** [comment]
I have an alternative proposal, and I'm sorry it's taken my time to write this down. My proposal is this: let's remove the thread pool as a need for bulk processor. We can do this in the following way:
 - the thread pool is needed for scheduling: for this we can either create a scheduling thread pool executor in the processor, and close on close or have the user pass one in (I favor the former, the latter is closer to today)
 - we use the thread pool to preserve the context for situations like reindex and it's friends; I think we can do this (thanks @nik9000 for helping with this one): allow a Function<Runnable, Runnable> to be passed in so then reindex and friends can pass a context preserving transformation and everyone else can us Function.identity

That would remove the need for this change.

**Comment by nik9000** [comment]
> we use the thread pool to preserve the context for situations like reindex and its friends; I think we can do this (thanks @nik9000 for helping with this one): allow a Function<Runnable, Runnable> to be passed in so then reindex and friends can pass a context preserving transformation and everyone else can us Function.identity

Now that you've written this I wonder if it'd be better instead to do
`Retry(BiFunction<Long, Runnable> delay)` and internally we delegate that appropriately to ThreadPool and others can delegate to their own delay implementation.

**Comment by javanna** [comment]
Closing, I will revisit based on Jason's last comment.

**Pull request closed without merging**